File: scripts/vectors/gvvectors/constructs/children.py
"""The children record (records.md#5): its canonical JSON, the owner's
signature, and opening the sealed record (in the open direction)."""

from .. import proto
from ..prims import b64url, ed_sign, seed, sealed_box, sealed_box_open


def build(f):
    root = bytes(range(32))
    owner = proto.Owner(proto.node_key(root, "acme/prod"))
    other = proto.Owner(proto.node_key(root, "acme/dev"))
    g = proto.Generation("")
    kids = [
        {"seg": "staging", "created_at": 1_757_500_200, "mode": {"kind": "derived"}},
        {"seg": "dev", "created_at": 1_757_500_100, "mode": {"kind": "derived"}},
        {"seg": "eu", "created_at": 1_757_500_300, "mode": {"kind": "sealed", "key_ct": b64url(seed("sealed child"))}},
    ]
    canonical = proto.children_plaintext(kids)
    f.add("encode", "records.md#5.1", "three children, sorted by segment, compact JSON",
          {"children": kids}, outputs={"plaintext": canonical})
    f.add("encode", "records.md#5.1", "no children", {"children": []},
          outputs={"plaintext": proto.children_plaintext([])})
    f.add("parse", "records.md#5.1", "a record in any order parses to the canonical one",
          {"plaintext": '{"v":1,"children":' + str(kids).replace("'", '"') + "}"}, outputs={"plaintext": canonical})
    for desc, text, expect in [
        ("version 0", '{"v":0,"children":[]}', "unknown_version"),
        ("version 2", '{"v":2,"children":[]}', "unknown_version"),
        ("a segment listed twice",
         '{"v":1,"children":[{"seg":"a","created_at":1,"mode":{"kind":"derived"}},'
         '{"seg":"a","created_at":2,"mode":{"kind":"derived"}}]}', "bad_encoding"),
        ("an unknown field", '{"v":1,"children":[],"server":"x"}', "bad_encoding"),
        ("an invalid segment", '{"v":1,"children":[{"seg":"Bad","created_at":1,"mode":{"kind":"derived"}}]}',
         "bad_encoding"),
        ("an unknown entry mode", '{"v":1,"children":[{"seg":"a","created_at":1,"mode":{"kind":"linked"}}]}',
         "bad_encoding"),
        ("not JSON", "children", "bad_encoding"),
    ]:
        f.add("parse", "records.md#5.1", f"refused: {desc}", {"plaintext": text}, expect)

    n = [0]

    def seal(to, text: str) -> bytes:
        n[0] += 1
        ct = sealed_box(to.box_pub, text.encode(), seed(f"children esk {n[0]}"))
        assert sealed_box_open(to.box_secret, ct) == text.encode()
        return ct

    ct = seal(owner, canonical)
    sig = owner.sign(proto.children_input(owner.vault_id, 4, ct))
    f.add("sign", "records.md#5.2", "the owner signs frame(gv/v1/children, vault_id ‖ version ‖ SHA-256(ct))",
          {"owner_node_key": owner.node.hex(), "version": 4, "ct": ct.hex()},
          outputs={"signing_input": proto.children_input(owner.vault_id, 4, ct).hex(), "sig": sig.hex()})

    def open_case(desc, version, ct_, sig_, expect="success", outputs=None):
        f.add("open", "records.md#5.3", desc,
              {"owner_node_key": owner.node.hex(), "version": version, "ct": ct_.hex(), "sig": sig_.hex()},
              expect, outputs)

    open_case("the owner opens its children record", 4, ct, sig, outputs={"plaintext": canonical})
    open_case("signed by the secret writer key (what an append token holds)", 4, ct,
              ed_sign(g.secret_writer, proto.children_input(owner.vault_id, 4, ct)), "bad_signature")
    open_case("served as another version", 5, ct, sig, "bad_signature")
    moved = seal(owner, canonical)
    open_case("signed for another vault", 4, moved,
              other.sign(proto.children_input(owner.vault_id, 4, moved)), "bad_signature")
    theirs = seal(other, canonical)
    open_case("sealed to another owner, signed by this one", 4, theirs,
              owner.sign(proto.children_input(owner.vault_id, 4, theirs)), "decrypt_failed")
    old = seal(owner, '{"v":2,"children":[]}')
    open_case("a version 2 record, correctly sealed and signed", 4, old,
              owner.sign(proto.children_input(owner.vault_id, 4, old)), "unknown_version")
    dup = seal(owner, '{"v":1,"children":[{"seg":"a","created_at":1,"mode":{"kind":"derived"}},'
                      '{"seg":"a","created_at":2,"mode":{"kind":"derived"}}]}')
    open_case("a segment listed twice, correctly sealed and signed", 4, dup,
              owner.sign(proto.children_input(owner.vault_id, 4, dup)), "bad_encoding")

File: crates/galata-vault-proto/src/children.rs
//! The children record: how a node knows its children.
//!
//! It is an **owner-only** record at its own endpoint
//! (`GET` and `PUT /v1/vault/children`), not a secret:
//! - sealed to the node's owner box key, so no token can read it;
//! - signed by the owner signing key, so no token, and not the server, can
//!   forge it;
//! - versioned, with the same preconditions as any record.
//!
//! Decrypted, it lists each child's segment and how to obtain the child's key:
//!
//! * `derived`: recompute it from this node's key (one-way HKDF). Carries
//!   no key material.
//! * `sealed`: a random key (after a rekey re-rooted the child), sealed to
//!   this node's owner box key as bundle kind 7.

use serde::{Deserialize, Serialize};

use crate::frame::{frame, label};
use crate::ids::{B64, Hash32, Key32, Sig64, VaultId};
use crate::integrity::{IntegrityError, verify_ed25519};
use crate::path::Segment;

/// Secret names starting with this are reserved; users cannot create them.
pub const RESERVED_PREFIX: &str = "gv:";
pub const CHILDREN_RECORD_VERSION: u32 = 1;

pub fn is_reserved_name(name: &str) -> bool {
    name.starts_with(RESERVED_PREFIX)
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "lowercase", deny_unknown_fields)]
#[non_exhaustive]
pub enum ChildMode {
    Derived,
    Sealed { key_ct: B64 },
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
#[non_exhaustive]
pub struct ChildEntry {
    pub seg: Segment,
    pub created_at: i64,
    pub mode: ChildMode,
}

impl ChildEntry {
    pub fn new(seg: Segment, created_at: i64, mode: ChildMode) -> ChildEntry {
        ChildEntry {
            seg,
            created_at,
            mode,
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct ChildrenRecord {
    pub v: u32,
    children: Vec<ChildEntry>,
}

#[derive(Debug, Clone, PartialEq, Eq, thiserror::Error)]
#[non_exhaustive]
pub enum ChildrenError {
    #[error("children record version {0} is not supported")]
    Version(u32),
    #[error("children record lists {0:?} more than once")]
    Duplicate(String),
    #[error("{0:?} is already a child of this node")]
    Exists(String),
    #[error("{0:?} is not a child of this node")]
    Missing(String),
    #[error("children record is not valid JSON: {0}")]
    Json(String),
}

impl Default for ChildrenRecord {
    fn default() -> Self {
        ChildrenRecord::new()
    }
}

impl ChildrenRecord {
    pub fn new() -> ChildrenRecord {
        ChildrenRecord {
            v: CHILDREN_RECORD_VERSION,
            children: Vec::new(),
        }
    }

    pub fn parse(bytes: &[u8]) -> Result<ChildrenRecord, ChildrenError> {
        let mut record: ChildrenRecord =
            serde_json::from_slice(bytes).map_err(|e| ChildrenError::Json(e.to_string()))?;
        if record.v != CHILDREN_RECORD_VERSION {
            return Err(ChildrenError::Version(record.v));
        }
        record.children.sort_by(|a, b| a.seg.cmp(&b.seg));
        if let Some(w) = record.children.windows(2).find(|w| w[0].seg == w[1].seg) {
            return Err(ChildrenError::Duplicate(w[0].seg.to_string()));
        }
        Ok(record)
    }

    /// Canonical bytes: children sorted by segment.
    pub fn to_bytes(&self) -> Result<Vec<u8>, ChildrenError> {
        serde_json::to_vec(self).map_err(|e| ChildrenError::Json(e.to_string()))
    }

    pub fn children(&self) -> &[ChildEntry] {
        &self.children
    }

    pub fn get(&self, seg: &Segment) -> Option<&ChildEntry> {
        self.children.iter().find(|c| &c.seg == seg)
    }

    pub fn insert(&mut self, entry: ChildEntry) -> Result<(), ChildrenError> {
        match self.children.binary_search_by(|c| c.seg.cmp(&entry.seg)) {
            Ok(_) => Err(ChildrenError::Exists(entry.seg.to_string())),
            Err(at) => {
                self.children.insert(at, entry);
                Ok(())
            }
        }
    }

    /// Replace an existing child's mode (a rekey turns `derived` into `sealed`).
    pub fn set_mode(&mut self, seg: &Segment, mode: ChildMode) -> Result<(), ChildrenError> {
        let entry = self
            .children
            .iter_mut()
            .find(|c| &c.seg == seg)
            .ok_or_else(|| ChildrenError::Missing(seg.to_string()))?;
        entry.mode = mode;
        Ok(())
    }

    pub fn remove(&mut self, seg: &Segment) -> Result<ChildEntry, ChildrenError> {
        let at = self
            .children
            .iter()
            .position(|c| &c.seg == seg)
            .ok_or_else(|| ChildrenError::Missing(seg.to_string()))?;
        Ok(self.children.remove(at))
    }
}

/// The owner's signature input for a children record:
/// `frame("gv/v1/children", vault_id(16) ‖ version(8) ‖ SHA-256(ct))`.
pub fn children_signing_input(vault_id: &VaultId, version: u64, ct: &[u8]) -> Vec<u8> {
    frame(
        label::CHILDREN,
        &[&vault_id.0, &version.to_be_bytes(), &Hash32::sha256(ct).0],
    )
}

/// The children record as it is stored and served: its version, the sealed
/// record and the owner's signature.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[non_exhaustive]
pub struct ChildrenBlob {
    pub version: u64,
    pub ct: B64,
    pub sig: Sig64,
}

impl ChildrenBlob {
    pub fn new(version: u64, ct: B64, sig: Sig64) -> ChildrenBlob {
        ChildrenBlob { version, ct, sig }
    }

    /// Check the owner's signature. Clients refuse anything that fails.
    pub fn verify(&self, vault_id: &VaultId, owner_sign_pub: &Key32) -> Result<(), IntegrityError> {
        verify_ed25519(
            owner_sign_pub,
            &children_signing_input(vault_id, self.version, &self.ct.0),
            &self.sig,
            "children record",
        )
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn seg(s: &str) -> Segment {
        Segment::new(s).unwrap()
    }

    fn entry(s: &str) -> ChildEntry {
        ChildEntry {
            seg: seg(s),
            created_at: 1,
            mode: ChildMode::Derived,
        }
    }

    #[test]
    fn roundtrip_is_sorted_and_canonical() {
        let mut r = ChildrenRecord::new();
        r.insert(entry("staging")).unwrap();
        r.insert(entry("dev")).unwrap();
        r.insert(ChildEntry {
            seg: seg("prod"),
            created_at: 2,
            mode: ChildMode::Sealed {
                key_ct: B64(vec![1, 2, 3]),
            },
        })
        .unwrap();
        let names: Vec<_> = r.children().iter().map(|c| c.seg.as_str()).collect();
        assert_eq!(names, ["dev", "prod", "staging"]);
        let bytes = r.to_bytes().unwrap();
        assert_eq!(ChildrenRecord::parse(&bytes).unwrap(), r);
        let text = String::from_utf8(bytes).unwrap();
        assert!(text.contains(r#""mode":{"kind":"derived"}"#));
        assert!(text.contains(r#""kind":"sealed","key_ct":"AQID""#));
    }

    #[test]
    fn duplicates_versions_and_unknown_fields_are_refused() {
        let mut r = ChildrenRecord::new();
        r.insert(entry("dev")).unwrap();
        assert_eq!(
            r.insert(entry("dev")),
            Err(ChildrenError::Exists("dev".into()))
        );
        let dup = br#"{"v":1,"children":[{"seg":"a","created_at":1,"mode":{"kind":"derived"}},{"seg":"a","created_at":2,"mode":{"kind":"derived"}}]}"#;
        assert_eq!(
            ChildrenRecord::parse(dup),
            Err(ChildrenError::Duplicate("a".into()))
        );
        assert_eq!(
            ChildrenRecord::parse(br#"{"v":2,"children":[]}"#),
            Err(ChildrenError::Version(2))
        );
        assert!(ChildrenRecord::parse(br#"{"v":1,"children":[],"server":"x"}"#).is_err());
        assert!(
            ChildrenRecord::parse(
                br#"{"v":1,"children":[{"seg":"Bad","created_at":1,"mode":{"kind":"derived"}}]}"#
            )
            .is_err()
        );
    }

    #[test]
    fn detach_and_remove() {
        let mut r = ChildrenRecord::new();
        r.insert(entry("prod")).unwrap();
        r.set_mode(
            &seg("prod"),
            ChildMode::Sealed {
                key_ct: B64(vec![9]),
            },
        )
        .unwrap();
        assert!(matches!(
            r.get(&seg("prod")).unwrap().mode,
            ChildMode::Sealed { .. }
        ));
        assert!(r.remove(&seg("prod")).is_ok());
        assert_eq!(
            r.remove(&seg("prod")),
            Err(ChildrenError::Missing("prod".into()))
        );
    }

    #[test]
    fn reserved_names() {
        assert!(is_reserved_name("gv:children"));
        assert!(is_reserved_name("gv:anything"));
        assert!(!is_reserved_name("GV_TOKEN"));
        assert!(!is_reserved_name("DATABASE_URL"));
    }

    #[test]
    fn the_signing_input_binds_vault_version_and_ciphertext() {
        let base = children_signing_input(&VaultId([1; 16]), 3, b"ct");
        assert!(base.starts_with(b"gv/v1/children\0"));
        assert_ne!(base, children_signing_input(&VaultId([2; 16]), 3, b"ct"));
        assert_ne!(base, children_signing_input(&VaultId([1; 16]), 4, b"ct"));
        assert_ne!(base, children_signing_input(&VaultId([1; 16]), 3, b"cu"));
    }
}
